Implement initial BindLeft and MonadLeft

JordanMartinez opened the pull request:
Fixes #12

WIP

Things to do:
- [x] Implement qualified do for `ldiscard`, `lbind`, `lmap`, and `lapply`

JordanMartinez commented on src/Control/MonadLeft/Qualified.purs:
```diff
@@ -0,0 +1,40 @@
+-- | import Control.MonadLeft.Qualified as MonadLeft
+-- |
+-- | foo :: Either Int String -> Either String String
+-- | foo comp = MonadLeft.ado
```
Should the convention be `Flip.do` rather than `MonadLeft.do`?

kl0tl replied:
`MonadLeft.do` seems more meaningful to me than `Flip.do`, but writing `MonadLeft.ado` for a type with only an `ApplicativeLeft` instance and no `MonadLeft` is odd.

JordanMartinez replied:
I read `MonadLeft.do` / `MonadLeft.ado` less of a "hey, this thing is a monad!" and more of a "hey, we're using the `MonadLeft` type class hierarchy here. That's how I read `Ix.do`, `React.do`, and `Hooks.do`.

JordanMartinez commented:
What advantage would there be for defining an instance of `MonadLeft` and its superclasses instances for `Flip` if we have qualified do?

kl0tl commented on src/Control/MonadLeft/Qualified.purs:
````diff
@@ -0,0 +1,40 @@
+-- |   b <- comp
+-- |   lpure $ show $ a + b
+-- | ```
+module Control.MonadLeft.Qualified where
````
Don’t we also need to export `pure` for Applicative do notation?

JordanMartinez replied:
I considered `Left.do`, but wasn't sure whether that would clash with the `Left` constructor for `Either`. `Flip` has some similarity to the `Flip` newtype, but it won't prevent clashing either. `MonadLeft` seems a bit much to type when other usages (e.g. `Ix`, `React`, `Hooks`, etc.) are shorter.

What about `BiLeft.do`? `Bi` for bifunctor and `Left` to indicate which parameter it's working on? I'm not sure whether every MonadLeft will have also be a Bifunctor... Maybe `MLeft` as a shortened version of `MonadLeft`?

JordanMartinez replied:
> Don’t we also need to export pure for Applicative do notation?

I don't believe so. Any usage of `pure` will likely be prefixed with the qualified do name. For example:
```purescript
main = Ix.do
  foo <- Ix.pure "foo"
  -- ...
```

kl0tl reviewed (commented):
While `lapply`, `lpure`, `lbind` and `ldiscard` are consistent with `lmap`, `llift*` is awkward and `lifM` reads like a typo. This is consistent with the prefixes from bifunctors and indexed-monad though, so perhaps this ship has already sailed.

JordanMartinez commented:
> This is consistent with the prefixes from bifunctors and indexed-monad though, so perhaps this ship has already sailed.

We could always break convention here now and update bifunctors and indexed-monad to use a `*Left` suffix in a future breaking changes release. In this case, I'd rather have something readable than something short.

JordanMartinez commented on src/Control/ApplyLeft.purs:
```diff
@@ -0,0 +1,36 @@
+import Data.Bifunctor (class Bifunctor, lmap)
+
+-- | Same as `Apply` but works on the left parameter in a Bifunctor.
+class Bifunctor w <= ApplyLeft w where
```
Pretty sure I copied and pasted this from somewhere and then updated it. So, it should be updated to `m`.

JordanMartinez commented on src/Control/BindLeft.purs:
```diff
@@ -0,0 +1,30 @@
+  ldiscard :: forall k (m :: Type -> k -> Type) b r. BindLeft m => m a r -> (a -> m b r) -> m b r
+
+instance discardLeftUnit :: DiscardLeft Unit where
+  ldiscard ma aToMB = lbind ma aToMB
```
Fixed in latest commits

JordanMartinez commented on src/Control/BindLeft.purs:
```diff
@@ -0,0 +1,30 @@
+lcomposeKleisli aToMB bToMC a = lbind (aToMB a) bToMC
+
+lcomposeKleisliFlipped :: forall m a b c r. BindLeft m => (b -> m c r) -> (a -> m b r) -> a -> m c r
+lcomposeKleisliFlipped bToMC aToMB a = lbind (aToMB a) bToMC
```
Fixed in latest commits

JordanMartinez commented on src/Control/ApplicativeLeft.purs:
```diff
@@ -0,0 +1,18 @@
+import Control.ApplyLeft (class ApplyLeft, lapply)
+
+-- | Same as `Applicative` but works on the left parameter in a Bifunctor.
+class ApplyLeft w <= ApplicativeLeft w where
```
Fixed in latest commits

JordanMartinez commented:
CI now builds.

JordanMartinez commented:
The only thing blocking this PR is deciding on the naming convention for referencing this qualified do. I think `BiLeft` should be used as it's short for `bifunctor` and the `left` parameter used in the bifunctor.

JordanMartinez commented:
I moved all the modules implemented here into a `Control.Bifunctor` module. This is so we reduce possible breaking changes if we ever want to implement a `MonadLeft`-like type class hierarchy for trifunctors. 

I've also updated the naming convention for the qualified do to use `BiLeft` which distinguishes it from the `Left` constructor of Either and ties it to the underlying bifunctor type's left type parameter.

Can I get an approval on this?

hdgarrood commented:
> This is so we reduce possible breaking changes if we ever want to implement a MonadLeft-like type class hierarchy for trifunctors.

This feels like a case of YAGNI to me. It's hard for me to see us ever wanting to do that in core; even this version of MonadLeft has only generated minimal interest.

hdgarrood commented:
I don't have a strong view on whether MonadLeft should be in core or not, but I think the argument that it should be here so that we can provide `Monad Flip` makes sense to me. It's just that I think it probably is safe to say that a version of MonadLeft which works on the first of _three_ type variables is not useful often enough to deserve to be in core, and therefore I'd suggest not designing this library in such a way as to make space to do that in the future.

JordanMartinez commented:
Ah... That does make sense.

JordanMartinez commented:
Closing as this hasn't received an approval or further feedback.
